cirq/optimizers/controlled_gate_decomposer.py

```diff
@@ -0,0 +1,284 @@
+            ops.ZPowGate(exponent=-0.25).on(c1),
+            ops.CNOT.on(c0, c1),
+        ]
+
+    def _ccnot_congruent(self, c0: 'cirq.Qid', c1: 'cirq.Qid',
+                         target: 'cirq.Qid'):
+        """Implements 'congruent' CCNOT.
+
+        This is almost like CCNOT, but multiplies 4th qubit by -1.
+        See lemma 6.2 in [1]."""
+        return [
+            ops.ry(-np.pi / 4).on(target),
+            ops.CNOT(c1, target),
+            ops.ry(-np.pi / 4).on(target),
+            ops.CNOT(c0, target),
+            ops.ry(np.pi / 4).on(target),
+            ops.CNOT(c1, target),
+            ops.ry(np.pi / 4).on(target),
+        ]
+
+    def _ccnot(self,
+               c0: 'cirq.Qid',
+               c1: 'cirq.Qid',
+               target: 'cirq.Qid',
+               congruent: bool = False):
+        """Implements CCNOT gate.
+
+        If `congruent=True`, implements "congruent" CCNOT gate, which differs
+        from CCNOT in that it multiplies 4th qubit by -1, but takes fewer gates
+        to implement.
+        """
+        if self.allow_toffoli:
+            return [ops.CCNOT.on(c0, c1, target)]
+        elif congruent:
+            return self._ccnot_congruent(c0, c1, target)
+        else:
+            return self._ccnot_true(c0, c1, target)
+
+    def decompose_x(self,
+                    controls: List['cirq.Qid'],
+                    target: 'cirq.Qid',
+                    free_qubits: List['cirq.Qid'] = None):
+        """Implements action of multi-controlled Pauli X gate .
+
+        Result is guaranteed to consist exclusively of 1-qubit, CNOT and CCNOT
+        gates.
+
+        If `free_qubits` has at least 1 element, result has length
+        O(len(controls)).
+
+        Args:
+            controls - control qubits. Can be empty.
+            targets - target qubits.
+            free_qubits - qubits which are neither controlled nor target. Can
+                be modified by algorithm but will end up in inital state.
+        """
+        if free_qubits is None:
+            free_qubits = []
+        m = len(controls)
+        if m == 0:
+            return [ops.X.on(target)]
+        elif m == 1:
+            return [ops.CNOT.on(controls[0], target)]
+        elif m == 2:
+            return self._ccnot(controls[0], controls[1], target)
+
+        m = len(controls)
+        n = m + 1 + len(free_qubits)
+        if (n >= 2 * m - 1) and (m >= 3):
+            # See [1], Lemma 7.2.
+            gates1 = [
+                self._ccnot(controls[m - 2 - i],
+                            free_qubits[m - 4 - i],
+                            free_qubits[m - 3 - i],
+                            congruent=True) for i in range(m - 3)
+            ]
+            gates2 = self._ccnot(controls[0],
+                                 controls[1],
+                                 free_qubits[0],
+                                 congruent=True)
+            gates3 = _flatten(gates1) + gates2 + _flatten(gates1[::-1])
```

Comment:
`gates3 = [*gates1, gates2, *gates1[::-1]]`

Reply:
No, this doesn't work. This would work if gates1 would be List of gates. But it's list of lists of gates.